servers/drivers/time/rtc/main.c

```diff
@@ -0,0 +1,90 @@
+#include <resea/printf.h>
+#include <vprintf.h>
+#include <driver/io.h>
+#include <resea/ipc.h>
+#include <resea/malloc.h>
+#include <string.h>
+#include "rtc.h"
+
+static io_t rtc_io;
+
+int is_update_in_progress(void) {
+	io_write8(rtc_io, 0, 10);
+    char RTC_UIP_FLAG = io_read8(rtc_io, 1);
+	return RTC_UIP_FLAG & 0x80;
+}
+
+uint8_t read_cmos_register(int reg_index) {
+    io_write8(rtc_io, 0, reg_index);
+    return (uint8_t) io_read8(rtc_io, 1);
+}
+
+uint8_t bcd_to_binary(uint8_t bcd) {
+    return (bcd & 0x0F) + ((bcd >> 4) * 10);
+}
+
+void read_datetime(uint8_t* year, uint8_t* month, uint8_t* day, uint8_t* hour, uint8_t* minute, uint8_t* second, uint8_t* day_of_week) {
```

Comment:
How about defining a struct for its return value?

```c
struct datetime {
    uint8_t year;
    uint8_t month;
    ...
};

void read_datetime(struct datetime *datetime) {
```

Reply:
Yeah, this is something I was debating. Should I define the struct for use in this driver only? Or should I define it as a global type in `types.h`? That would also allow using the same type in message passing. And it will be also useful when converting time from UNIX-like timestamp.

Reply:
Please define it for this driver only for now. Let's consider moving it into the global scope when we actually need to do. 

JFYI, the IPC stub generator does not yet support `struct` types in the message payload.

Reply:
Yeah, I did notice that the generator didn't support that. That's why I didn't go with struct. What you proposed seems the best for now.